solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,316 @@
+    {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+
+        // TODO: Validate when it was revealed. The deposit must be revealed
+        //       early enough to the Bridge to pass the Bridge's validation.
+        //       It may happen that none of the Minters request for an
+        //       optimistic mint of a deposit early and one of them do it just
+        //       before the Bitcoin refund unlocks. It should not be possible
+        //       to request an optimistic mint for such a deposit because
+        //       there is no guarantee the wallet will be able to sweep it.
+
+        require(deposit.revealedAt != 0, "The deposit has not been revealed");
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+        require(deposit.vault == address(this), "Unexpected vault address");
+
+        /* solhint-disable-next-line not-rely-on-time */
+        pendingOptimisticMints[depositKey] = block.timestamp;
+
+        emit OptimisticMintingRequested(
+            msg.sender,
+            deposit.depositor,
+            deposit.amount,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Allows a Minter to finalize previously requested optimistic
+    ///         minting. The following conditions must be met:
+    ///         - The optimistic minting has been requested for the given
+    ///           deposit.
+    ///         - The deposit has not been swept yet.
+    ///         - At least OPTIMISTIC_MINTING_DELAY passed since the optimistic
+    ///           minting was requested for the given deposit.
+    ///         - The optimistic minting has not been finalized earlier for the
+    ///           given deposit.
+    ///         - The optimistic minting request for the given deposit has not
+    ///           been canceled by a Guardian.
+    ///         This function mints TBTC and increases pendingOptimisticMints
+    ///         for the given depositor. The finalized optimistic minting
+    ///         request is removed from the contract.
+    function finalizeOptimisticMint(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) external onlyMinter {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+
+        uint256 requestedAt = pendingOptimisticMints[depositKey];
+        require(
+            requestedAt != 0,
+            "Optimistic minting not requested or already finalized"
+        );
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            block.timestamp - requestedAt > OPTIMISTIC_MINTING_DELAY,
+            "Optimistic minting delay has not passed yet"
+        );
+
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
```

Comment:
There's this comment https://github.com/keep-network/tbtc-v2/blob/main/solidity/contracts/bridge/Bridge.sol#L1504-L1506 about a `deposits` map containing valid and invalid deposits that a wallet is responsible to validate them. Are we going to use the same validation here?

Reply:
Not in the contract but Minters and Guardians are supposed to perform this validation, yes. We do not know if the deposit revealed to Ethereum happened at all on Bitcoin.